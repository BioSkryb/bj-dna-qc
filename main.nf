nextflow.enable.dsl=2

include { printHeader; helpMessage } from './help' params ( params )
include { PRESEQ_WF } from './workflow/preseq.nf' params ( params )


if ( params.help ) {
    helpMessage()
    exit 0
}

params.reference     = params.genomes [ params.genome ] [ 'reference' ]
params.intervals     = params.genomes [ params.genome ] [ 'base_metrics_intervals' ]
params.ginko_ref_dir = params.genomes [ params.genome ] [ "${params.bin_size}" ] [ 'ginko_ref_dir' ]
params.krakendb      = params.genomes [ params.genome ] [ 'kraken2_db' ]

workflow {
    
    printHeader()
    
    // Check if publishDir specified
    
    if ( params.publish_dir == "") {
        exit 1, "ERROR: publish_dir is not defined.\nPlease add --publish_dir s3://<bucket_name>/<project_name> to specify where the pipeline outputs will be stored."
    }
    
    if ( params.reads ) {
        
        ch_reads = Channel.fromFilePairs( params.reads , size: -1 , checkExists: true )    
        ch_reads.ifEmpty{ exit 1, "ERROR: cannot find any fastq files matching the pattern: ${params.reads}\nMake sure that the input file exists!" }

    } else if ( params.input_csv ) {
        
        ch_reads = Channel.fromPath( params.input_csv ).splitCsv( header:true )
                            .map { row -> [ row.biosampleName, [ row.read1, row.read2 ] ] }
        ch_reads.ifEmpty{ exit 1, "ERROR: Input csv file is empty." }
    }
    
    

    ch_reads.view()
    ch_dummy_file = Channel.fromPath("$projectDir/assets/dummy_file.txt", checkIfExists: true).collect()
    ch_dummy_file2 = Channel.fromPath("$projectDir/assets/dummy_file2.txt", checkIfExists: true).collect()
    
    ch_reference =  Channel.fromPath(params.reference)
    ch_intervals =  Channel.fromPath(params.intervals)
    
    ch_binref = Channel.fromPath(params.ginko_ref_dir + "variable_" + params.bin_size + "_" + params.ginko_readlen + "_bwa")
    ch_gcref = Channel.fromPath(params.ginko_ref_dir + "GC_variable_" + params.bin_size + "_" + params.ginko_readlen + "_bwa")
    ch_boundsref_file = Channel.fromPath(params.ginko_ref_dir + "bounds_variable_" + params.bin_size + "_" + params.ginko_readlen + "_bwa")

       
    PRESEQ_WF( 
                params.publish_dir,
                params.enable_publish,
                params.disable_publish,
                ch_reads,
                params.input_csv,
                ch_dummy_file,
                ch_dummy_file2,
                params.mode,
                params.multiqc_config,
                params.project,
                params.bin_size,
                params.min_ploidy,
                params.max_ploidy,
                params.min_bin_width,
                params.is_haplotype,
                ch_reference,
                ch_intervals,
                ch_binref,
                ch_gcref,
                ch_boundsref_file,
                params.seqtk_sample_seed
             )
    
    
}

// OnComplete
workflow.onComplete{
    println( "\nPipeline completed successfully.\n\n" )
}

// OnError
workflow.onError{
    println( "\nPipeline failed.\n\n" )
}
