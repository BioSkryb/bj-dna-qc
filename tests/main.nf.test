nextflow_pipeline {

    name "Test Workflow main.nf"
    script "main.nf"

    test("preseq_test") {
        when {
            params {
                // define parameters here. Example: 
                publish_dir = "${outputDir}/results"
                timestamp = "test"
                architecture = "x86"
                input_csv = "$baseDir/tests/data/inputs/input.csv"
                skip_kraken = false
                skip_qualimap = false
                skip_mapd = false
                skip_fastqc = false
                skip_ginkgo = false
                max_memory = "14.GB"
                max_cpus = "4"
                max_time = "2.h"
            }
        }

        then {
            assertAll(
                
                // Check if the workflow was successful
                { assert workflow.success },

                // Verify existence of the multiqc report HTML file
                { 
                    assert new File("${outputDir}/results_test/multiqc/multiqc_report.html").exists()
                },

                // Check for a match in the all metrics MQC text file
                { 
                    assert snapshot(path("${outputDir}/results_test/secondary_analyses/metrics/nf-preseq-pipeline_all_metrics_mqc.txt"))
                        .match("all_metrics_mqc")
                },

                // Check for a match in the selected metrics MQC text file
                { 
                    assert snapshot(path("${outputDir}/results_test/secondary_analyses/metrics/nf-preseq-pipeline_selected_metrics_mqc.txt"))
                        .match("selected_metrics_mqc")
                },

                // Verify existence of the fastp JSON file
                { 
                    assert new File("${outputDir}/results_test/primary_analyses/metrics/DNAQC-test2-1000reads/fastp/DNAQC-test2-1000reads_no_qc_fastp.json").exists()
                },

                // Check for a match in the kraken report text file
                { 
                    assert snapshot(path("${outputDir}/results_test/primary_analyses/metrics/DNAQC-test2-1000reads/kraken2/DNAQC-test2-1000reads_kraken2_report.txt"))
                        .match("kraken_report")
                },

                // Check for a match in the Ginkgo segment summary text file
                { 
                    assert snapshot(path("${outputDir}/results_test/tertiary_analyses/cnv_ginkgo/AllSample-GinkgoSegmentSummary.txt"))
                        .match("ginkgo_summary")
                }
            )
            
        }

    }

}
