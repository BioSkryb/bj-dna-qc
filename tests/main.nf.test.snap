{
    "selected_metrics_mqc": {
        "content": [
            "nf-preseq-pipeline_selected_metrics_mqc.txt:md5,80c5a494ea6e65f3e6b724c052b6ae55"
        ],
        "timestamp": "2024-11-19T15:47:51+0000"
    },
    "ginkgo_summary": {
        "content": [
            "AllSample-GinkgoSegmentSummary.txt:md5,7bef2918a7732bb65fcbab539b2e65a9"
        ],
        "timestamp": "2024-11-19T15:47:51+0000"
    },
    "all_metrics_mqc": {
        "content": [
            "nf-preseq-pipeline_all_metrics_mqc.txt:md5,ee42f826d2d214a065d7e9799c8244c7"
        ],
        "timestamp": "2024-11-19T15:47:51+0000"
    },
    "kraken_report": {
        "content": [
            "DNAQC-test2-1000reads_kraken2_report.txt:md5,52c96bce25dd27a5dec5284e32fab629"
        ],
        "timestamp": "2024-11-19T15:47:51+0000"
    }
}